File: Data Structures & Algorithms/subsets/submission-0.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    int n = 0;
    private void backtrack(int[] nums, ArrayList<Integer> arr, int i) {
        if(i < 0) {
            res.add(new ArrayList<>(arr));
            return;
        }
        arr.add(nums[i]);
        backtrack(nums, arr, i-1);
        arr.remove(arr.size() - 1);
        backtrack(nums, arr, i - 1);
    }
    public List<List<Integer>> subsets(int[] nums) {
        if(nums.length == 0) {
            return res;
        }
        n = nums.length;
        backtrack(nums, new ArrayList<Integer>() ,n-1);
        return res;
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(begin(nums), end(nums));
        int n = nums.size();
        
        for(int i=0; i<n-2; i++) {
            if(nums[i] > 0) break;
            if(i > 0 && nums[i] == nums[i-1]) continue;

            int j = i+1, k = n-1;
            while(j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == 0) {
                    res.push_back({ nums[i], nums[j], nums[k] });
                    while(j<k && nums[j] == nums[j+1]) j++;
                    while(j<k && nums[k] == nums[k-1]) k--;

                    j++, k--;
                }
                else if(sum > 0) k--;
                else j++;
            }
        }
            return res;
    }
};
